### PR TITLE
ZEPPELIN-186: Apache Phoenix interpreter

### PR DESCRIPTION
**Pull request by randerzander**
Include an interpreter for Apache Phoenix

Almost identical to Hive interpreter. Could/should generic JDBC interpreters be combined in future work?

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
Thanks for the contribution! LGTM.

I think having generic JDBC interpreter is good idea. especially if JDBC interpreter features something similar to [Dependency loader](http://zeppelin.incubator.apache.org/docs/interpreter/spark.html#dependencyloading), so it can download and load jdbc driver from maven repository.

**Comment by tzolov** [comment]
@randerzander, while working on the `postgresql` and `geode` i had to consider few details (credit to @martin-g for pointing them out) that you can find helpful as well . 

* You may want to define a maxResultRow to restrict the max number of result displayed or you risk to kill your zeppelin. In couple of occasions I've forgotten to set sql LIMIT  and managed to freeze the front end. 
* As a safeguard you can call the close() in your open method before opening the connection
* For testing use proper Mockito and MockRunner [PostgreSqlInterpreterTest.java](https://github.com/apache/incubator-zeppelin/blob/master/postgresql/src/test/java/org/apache/zeppelin/postgresql/PostgreSqlInterpreterTest.java)

**Comment by tzolov** [comment]
@Leemoonsoo, i've been thinking for implementing Dep loaded in the context of postgresql. My intention was to enable different psql drivers to be used but if we implement this then it will turn the psql interpreter into a generic JDBC interpreter (may have to change the name then). 
Is it possible to generalize the current Dependency implementation or reuse part of it?

**Comment by randerzander** [comment]
I suggest separating the Generic JDBC interpreter into its own feature as a separate effort, then picking which of the existing interpreters can be collapsed into it.

Thoughts?

**Comment by Leemoonsoo** [comment]
@randerzander what do you think about feedback from @tzolov ?

> * You may want to define a maxResultRow to restrict the max number of result displayed or you risk to kill your zeppelin. In couple of occasions I've forgotten to set sql LIMIT and managed to freeze the front end.
> * As a safeguard you can call the close() in your open method before opening the connection
> * For testing use proper Mockito and MockRunner PostgreSqlInterpreterTest.java

@tzolov generalizing Dependency implementation is definitely a direction (along with generalizing ZeppelinContext). We can think two different level of generalize.

1) generalize dependency downloading
 Generalize package [org.apache.zeppelin.spark.dep](https://github.com/apache/incubator-zeppelin/tree/master/spark/src/main/java/org/apache/zeppelin/spark/dep)
2) generalize dependency interpreter
 Generalize package [org.apache.zeppelin.spark.dep](https://github.com/apache/incubator-zeppelin/tree/master/spark/src/main/java/org/apache/zeppelin/spark/dep) and [DepInterpreter](https://github.com/apache/incubator-zeppelin/blob/master/spark/src/main/java/org/apache/zeppelin/spark/DepInterpreter.java)

**Comment by randerzander** [comment]
First, thanks for the review and feedback, @Leemoonsoo  and @tzolov -

I agree with the first 3, and will follow up with those changes.

However, I think we should keep the test as is for now. I will need to make a followup commit using PhoenixConnection and PhoenixStatement instead of the generic JDBC anyway- Phoenix has SQL syntax for dynamic columns that generic JDBC would reject as invalid.

**Comment by randerzander** [comment]
First two feedbacks incorporated.

@tzolov Having looked closer at your Mockito example, I think it may work. I'll follow up here ASAP.

**Comment by randerzander** [comment]
@tzolov @Leemoonsoo incorporated all feedbacks. Added ability to use upsert and DDL statements by switching to execute instead of executeQuery.

File: phoenix/src/main/java/org/apache/zeppelin/phoenix/PhoenixInterpreter.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+          //Append rows
+          int rowCount = 0;
+          while (res.next() && rowCount < getMaxResult()) {
+            row = clean(isExplain, res.getString(1));
+            for (int i = 2; i < md.getColumnCount() + 1; i++)
+              row += TAB + clean(isExplain, res.getString(i));
+            msg.append(row + NEWLINE);
+            rowCount++;
+          }
+          res.close();
```
Comment (tzolov) [inline phoenix/src/main/java/org/apache/zeppelin/phoenix/PhoenixInterpreter.java]: It is safer to close the ResultSet in the `final` block. Right now it seems it is not being closed in the case of false execute result.

File: phoenix/src/main/java/org/apache/zeppelin/phoenix/PhoenixInterpreter.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        else { // May have been upsert or DDL
+          msg.append(UPDATE_HEADER + NEWLINE +
+            "Rows affected: " + currentStatement.getUpdateCount()
+            + NEWLINE);
+        }
+
+        currentStatement.close();
```
Comment (tzolov) [inline phoenix/src/main/java/org/apache/zeppelin/phoenix/PhoenixInterpreter.java]: Same here. It is safer to close the currentStatement in the `final` block or you risk to leek resources.

**Comment by tzolov** [comment]
Apart from the comments above it looks good. 
i've forgotten to mention to use `execute` instead of `executeQuery` but i've figured it out :)

**Comment by tzolov** [comment]
@Leemoonsoo, i will start a new (dedicated) thread on the subject of generalizing and reusing the Dependency functionality.  Is there a JIRA ticket for it or shall i open new one? 

>@tzolov generalizing Dependency implementation is definitely a direction (along with generalizing ZeppelinContext). We can think two different level of generalize.
1) generalize dependency downloading
Generalize package org.apache.zeppelin.spark.dep
2) generalize dependency interpreter
Generalize package org.apache.zeppelin.spark.dep and DepInterpreter

**Comment by Leemoonsoo** [comment]
@tzolov I think you can open new one. Thanks!

**Comment by randerzander** [comment]
moving the result set and statement closes to the finally block broke upserts.. another commit coming once I figure that out.

**Comment by randerzander** [comment]
Thanks for your help @tzolov and @Leemoonsoo I think I've got it this time!

**Comment by randerzander** [comment]
any other feedback guys?

**Comment by youngwookim** [comment]
@randerzander  Thanks for your great work! Just out of curiosity, I believe it would be nice to have the default dependency from Apache release artifacts. i.e, 4.4.0-HBase-0.98 or 4.4.0-HBase-1.0 ...

The phoenix-core's version from Apache is a combination of Phoenix version and HBase version.

**Comment by randerzander** [comment]
Oops, good catch thanks!

**Comment by randerzander** [comment]
I used HBase 1.1 since that's the latest and deployed by default with the newest HDP.

Phoenix is releasing 4.5 soon- how does Zeppelin team prefer to manage version bumps?

**Comment by youngwookim** [comment]
HBase team suggest downloading 1.0.x[1] because it is the current stable release. IMO, 1.0 would be appropriate default for now. If there are incompatible changes which break building the interpreter, we may consider adding 'profiles' for versions in each release. What do you think? @randerzander

[1] http://apache.mirror.cdnetworks.com/hbase/

**Comment by randerzander** [comment]
rolled back to 1.0- seems to work fine with HBase 1.1 - Phoenix 4.4 deployments too. -Dphoenix.version build arguments will probably suffice.

**Comment by randerzander** [comment]
@Leemoonsoo @tzolov @youngwookim more feedback?

**Comment by Leemoonsoo** [comment]
Looks great!

**Comment by tzolov** [comment]
+1

**Comment by youngwookim** [comment]
+1

Thanks @randerzander !!

**Pull request closed without merging**